### DISCUSSION TITLE
Hide lines in document

### DISCUSSION THREAD

**Discussion by frasermarch** [microsoft/vscode-discussions#77] [Extension Development QnA] [answered]
I would like to make an extension that only shows certain lines of a document in the editor, hiding the rest, while allowing editing.

I've looked into folding, but I can't see a way to fold arbitrary lines. Any help would be really appreciated!

**Comment by alefragnani** [comment, marked as answer]
Hi @frasermarch ,

You should try the [FoldingRangeProvider](https://code.visualstudio.com/api/references/vscode-api#FoldingRangeProvider). Using it, you will be responsible to define the _range_ to be folded.

Hope this helps

**Reply by frasermarch** [reply]
Hey, thanks for this. I did see this but I'm unsure how to implement it. Do you have anything you could point me to?

**Reply by alefragnani** [reply]
Unfortunately not. Neither the extension-sample repo has a sample about this API 😞 

I did a small search in VS Code issues that could show samples about it, but didn’t find anything. On the other hand, I found some posts from the Python team trying out this API. And a few references in the VS Code source itself. I tried a sample code, the provider function was called, but no Folding appeared. I probably missed something.

**Reply by alefragnani** [reply]
My sample was something like, but with no success 😞 :

```ts
class TestFoldingRangeProvider  implements vscode.FoldingRangeProvider {
    onDidChangeFoldingRanges?: vscode.Event<void> | undefined;
    provideFoldingRanges(document: vscode.TextDocument, context: vscode.FoldingContext, token: vscode.CancellationToken): vscode.ProviderResult<vscode.FoldingRange[]> {
        return new Promise((resolve, reject) => {
            const foldingRange: vscode.FoldingRange = new vscode.FoldingRange(10, 20, vscode.FoldingRangeKind.Region)
            return [foldingRange];
        });
    }
}

...
    context.subscriptions.push(
        vscode.languages.registerFoldingRangeProvider(documentSelector, new TestFoldingRangeProvider())
    );
```

**Reply by frasermarch** [reply]
Thanks for trying this! I tried something similar but couldn't get it to work.

**Reply by r3m0t** [reply]
@alefragnani you need to call `resolve([foldingRange])` to send the value to the consumer of the Promise.
